Update cirros.tar.gz handling to be more agnostic of running within the image

tianon opened the pull request:
This is especially important for running the tests on platforms where the daemon isn't supported yet (Windows, Darwin, FreeBSD, etc).

tianon commented:
Talking this through with @ahmetalpbalkan because it won't work for Windows as-is. :cry:

ahmetalpbalkan commented:
LGMT. Jenkins test output https://jenkins.dockerproject.com/job/Windows-PRs/31/console

> `	docker_cli_events_test.go:223: import failed with errors: exit status 1, output: "Downloading from http://127.0.0.1:58850/cirros.tar.gz\ntime=\"2015-02-12T23:33:44Z\" level=\"fatal\" msg=\"Get http://127.0.0.1:58850/cirros.tar.gz: dial tcp 127.0.0.1:58850: connection refused\" \n"`

which means CIRROS_TAR path is plumbed to windows executable correctly. But since the test assumes daemon/cli are sitting on the same machine, serving file over a http server from cli-side doesn't work. I may fix later by using `fakeContext`.

tianon commented:
Why aren't we just using the Go equivalent of `docker import - < $CIRROS_TAR` here instead?

ahmetalpbalkan commented:
We might. Idk why test is designed like this in the first place. If there are no objections, I'm +1 for making it so.

tianon commented:
So, what if instead, we axed `cirros.tar.xz` completely?  We could do, essentially, `docker run -d busybox true`, `docker export <that container> | docker import - new-busybox` to test import appropriately without a second tarball or a second image.

ahmetalpbalkan commented:
@tianon yeah that's certainly better IMO. I'm not exactly sure how cirros is different than busybox. we don't even run it. it might be better to figure out how it ended up there.

tianon commented:
#8013 is where it ended up in the `Dockerfile` -- still digging for where it was added to the test initially

tianon commented:
#4930 -- looks like it was added as a "crap, we don't have a good test for import, let's add one!" :+1:

tianon commented:
Updating this now to just remove `cirros.tar.xz` completely now, since my test worked. :+1:

ahmetalpbalkan commented:
@tianon omg I'm now relieved. let's get rid of that.

tianon commented:
Updated:
```
commit db651a866d1c0d601b05e0ed7df86ccb71bdb423
Author: Tianon Gravi <admwiggin@gmail.com>
Date:   Thu Feb 12 14:40:34 2015 -0700

    Remove cirros.tar.gz completely
    
    Since `cirros.tar.gz` only existed to test `docker import`'s display and
    presence in `docker events`, we can instead just use `docker export`
    piped directly to `docker import` to achieve the same goal without
    another external dependency besides `busybox` (which we already have).
    
    While I was at it, I updated `TestImportDisplay` to also test that the
    imported image actually runs successfully as well (so we're testing the
    full import round-trip).
```

tianon committed:
Remove cirros.tar.gz completely

Since `cirros.tar.gz` only existed to test `docker import`'s display and
presence in `docker events`, we can instead just use `docker export`
piped directly to `docker import` to achieve the same goal without
another external dependency besides `busybox` (which we already have).

While I was at it, I updated `TestImportDisplay` to also test that the
imported image actually runs successfully as well (so we're testing the
full import round-trip).

Signed-off-by: Andrew "Tianon" Page <admwiggin@gmail.com>

ahmetalpbalkan commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -206,18 +206,26 @@ func TestEventsImagePull(t *testing.T) {
-	importCmd := exec.Command(dockerBinary, "import", fileURL, "cirros")
-	out, _, err := runCommandWithOutput(importCmd)
+	out, _, err = runCommandWithOutput(importCmd)
+	exportCmd.Wait() // release "exportCmd" resources
```
this returns err we might use?

tianon replied:
I thought about that, but in this case we don't really care if export
failed; only import matters here, and I only even bother invoking Wait to
let it release resources.

ahmetalpbalkan replied:
sounds fair, then.

ahmetalpbalkan replied:
actually I wondered if we need this at all. if the other end of pipe is broken, this cmd will also exit right away, correct?

tianon replied:
Right, but the description mentions that it also frees resources, so I was trying to be a responsible consumer.

ahmetalpbalkan commented on integration-cli/docker_cli_import_test.go:
```diff
@@ -1,32 +1,51 @@
+		t.Errorf("export failed with errors: %v", err)
+	}
+	out, _, err = runCommandWithOutput(importCmd)
+	exportCmd.Wait() // release "exportCmd" resources
```
this returns err which we might use?

ahmetalpbalkan commented:
@tianon you might want to update the PR title to reflect what actually happened.

thaJeztah commented:
@tianon https://github.com/docker/docker/pull/10750#issuecomment-74369775 :heart: a proper commit-message; this should
*really* become the norm (as the discussion above clearly proves).

SvenDowideit commented on integration-cli/docker_cli_events_test.go:
```diff
@@ -206,18 +206,26 @@ func TestEventsImagePull(t *testing.T) {
-	server, err := fileServer(map[string]string{
-		"/cirros.tar.gz": "/cirros.tar.gz",
-	})
+	exportCmd := exec.Command(dockerBinary, "export", cleanedContainerID)
```
can we validate that the export is a sane tar file? otherwise we might get to a point where some garbage export imports 'ok'

ahmetalpbalkan replied:
Doesn't "import" already implicitly validate export was ok?

SvenDowideit replied:
if there is an `import` test for this explicit image, then yes

tianon replied:
This test is for "docker events"; the test specifically for import does export, import, and then run to verify all the way through.

ahmetalpbalkan commented on integration-cli/docker_cli_import_test.go:
```diff
@@ -1,32 +1,51 @@
+
+	exportCmd := exec.Command(dockerBinary, "export", cleanedContainerID)
+	importCmd := exec.Command(dockerBinary, "import", "-")
+	importCmd.Stdin, err = exportCmd.StdoutPipe()
```
you might wanna check my `runCommandPipelineWithOutput` in #10806. it's the perfect tool for this job.

tianon replied:
There's a new function to save one line?

ahmetalpbalkan replied:
Heheh in this case 3 but yeah. There's a new function that can reduce 4 pipes to 1 variadic function :-)

ahmetalpbalkan replied:
It even does the "// release resources" for you :-)

tianon replied:
OK, in that case, I'm sold.  Not at a computer right now, but will update
today.

GordonTheTurtle commented:
These files are not properly gofmt'd:
integration-cli/docker_cli_import_test.go
Please reformat the above files using `gofmt -s -w` and ammend to the commit the result.

tianon commented:
Updated to take advantage of `runCommandPipelineWithOutput` appropriately, making the code even simpler.

ahmetalpbalkan commented:
:+1: :+1: :+1: :+1:

jfrazelle commented:
LGTM ping @cpuguy83
